src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -323,20 +323,36 @@ private void Init(string path, FileMode mode, FileAccess access, FileShare share
                 }
             }
 
-            // If preallocationSize has been provided for a creatable and writeable file
-            if (FileStreamHelpers.ShouldPreallocate(preallocationSize, access, mode, this))
+            if (preallocationSize > 0)
             {
-                int fallocateResult = Interop.Sys.PosixFAllocate(this, 0, preallocationSize);
-                if (fallocateResult != 0)
+                if (Interop.Sys.FAllocate(this, 0, preallocationSize) < 0)
                 {
-                    Dispose();
-                    Interop.Sys.Unlink(path!); // remove the file to mimic Windows behaviour (atomic operation)
-
-                    Debug.Assert(fallocateResult == -1 || fallocateResult == -2);
-                    throw new IOException(SR.Format(
-                        fallocateResult == -1 ? SR.IO_DiskFull_Path_AllocationSize : SR.IO_FileTooLarge_Path_AllocationSize,
-                        path,
-                        preallocationSize));
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+
+                    if (errorInfo.Error != Interop.Error.EOPNOTSUPP && // Not supported by filesystem.
+                        errorInfo.Error != Interop.Error.ENOSYS &&     // Not supported by kernel.
+                        errorInfo.Error != Interop.Error.ESPIPE &&     // Not a file.
+                        errorInfo.Error != Interop.Error.ENODEV)       // Not a file.
+                    {
+                        Dispose();
+                        Interop.Sys.Unlink(path!); // remove the file to mimic Windows behaviour (atomic operation)
```

Comment:
As I noted to Adam in his PR, this comment is misleading:
https://github.com/dotnet/runtime/pull/58726#discussion_r707364908
Is this unlinking behavior still a good idea?  If FAllocate fails, is it possible it left things in a state where lots of space was consumed, as was previously suggested of posix_fallocate?

Reply:
I suspect this will behave similar as `posix_fallocate`.
All of this happens as part of opening the file. If we've created the file and we throw, then we should delete it too.

Reply:
Before, this was only called on ENOSPC and EFBIG, now it can also be called when something like EINVAL is thrown. Is that OK?

Reply:
```suggestion
                        Interop.Sys.Unlink(path!); // when fallocate fails because there is not enough space, it usually consumes all available disk space (so the system is in a very bad shape), to mitigate that we remove the file.
```